source/host_resolver.c

```diff
@@ -1473,8 +1475,9 @@ static int default_resolve_host(
         if (aws_array_list_length(&callback_address_list)) {
             res(resolver, host_name, AWS_OP_SUCCESS, &callback_address_list, user_data);
         } else {
+            /* aws_last_error will be set by aws_host_address_copy or aws_array_list_push_back if something goes wrong.
+             */
             res(resolver, host_name, aws_last_error(), NULL, user_data);
```

Comment:
Maybe it's better to return error instead of firing the callback? The function failed synced, the return code is more proper to handle the synced failure.

Reply:
Makes sense, I think we should be consistent and just update aws-c-common guidelines to include callback rules or something.

Reply:
agree, but disagree

I agree with:
- we should be consistent
- we should update guidelines

But disagree on changing it as Dengke suggested.

In a perfect world, the callback would ALWAYS fire asynchronously. The synchronous failure via aws_raise_error() would be reserved for cases where it's impossible to do the lookup (the input arguments are invalid, thread can't be launched, etc). In cases where we failed to find any results, we'd still report that via callback, even if we knew the answer synchronously.

So let's take one of these paths:
- ALWAYS report results asynchronously. Change this code so we report success and failure asynchronously. We can schedule a task on the event_loop_group to do this. See comment `NOTE: we don't use el_group yet, but we will in the future`
- OR keep changes minimal and continue reporting both success and failure results synchronously

Reply:
Here's a long blog post explaining why callbacks that are SOMETIMES sync and SOMETIMES async are bad. You gotta always be one or always be the other:
https://blog.ometer.com/2011/07/24/callbacks-synchronous-and-asynchronous/

Reply:
added it to the guidelines:
https://github.com/awslabs/aws-c-common/pull/945